### PR TITLE
Feat user registration

### LINKED ISSUE #39: Feat: Register User (Vendors & Buyers)
## Description

As a vendor or buyer, I want to be able to register an account on the e-commerce platform so that I can access the respective features and functionalities.

The registration process should allow users to sign up as either a vendor or a buyer, and capture relevant information such as name, email, password, and any additional details required for each user type.

### Acceptance Criteria

- [x] Implement a user registration API endpoint
- [x] Validate user input (e.g., email, password, user type)
- [x] Check for existing users (email uniqueness)
- [x] Hash and store user passwords securely
- [x] Create user records in the database with appropriate user types (vendor or buyer)
- [x] Return appropriate success or error responses
- [x] Implement input sanitization

## Developer Notes

- Design the user registration data model and database schema
- Define the API endpoint and request/response payloads
- Implement input validation and sanitization measures
- Use a secure password hashing algorithm (e.g., bcrypt, argon2)
- Implement email user verification
- Write unit tests to cover different scenarios and edge cases
- Write clear API documentation for resulting endpoints

## Related Issues

- #40   
- #41
- #46  

## Definition of Done

- User registration API endpoint is implemented and functional
- User input validation and sanitization measures are in place
- Passwords are securely hashed and stored
- User records are created in the database with appropriate user types
- Appropriate success and error responses are returned
- Unit tests are written and passing
- Code adheres to coding standards and best practices
- All created endpoints are documented

## Implimentation plan

**1. Deliverables at the End of the Project**
Functionality:

-  Users can register with their details and get stored in the database.
 Technical Considerations:

- The implementation plan will consider the manager's provided document 

**2. Data Flow / Flowchart / Pseudocode / ERD Diagrams**
Tools:

- **Node.js:**
- Backend Server: Node.js will serve as the foundation for our backend server, handling user requests for password reset, generating tokens, and interacting with the database.

**- PostgreSQL:**
- Database: PostgreSQL will store your user data, including email addresses. 

**TypeORM:**

- Object-Relational Mapper (ORM): TypeORM acts as an ORM, simplifying interactions between your Node.js application and the PostgreSQL database. It allows you to write code using JavaScript classes and methods that map to database tables and queries.

**3. Schedule of Implementation**
Breakdown:
**Tuesday (April 30):**

- **Backend Setup:**

- Installed Node.js and set up your development environment.
- Installed TypeORM and configured your database connection (PostgreSQL).

- **Initial Routes and Controllers:**

 - Defined basic routes for handling user registration (e.g., /user/register).
 - Created controller functions to handle these routes.

**Wednesday (May 1st):**

- **Code Refactoring:**
- Reviewed and refactored the code based on your colleague's suggestions.
- Improved code structure, readability, and maintainability.
- Added comments to explain the code functionality.
- **Version Control (Git):**
- Committed your code changes to a version control system like Git.
- Pushed your code to a remote repository like GitHub.

**4. Dependencies/Foreseen Blockers**
Learning  typeOrm

**5. External Services or Libraries to be Used**

Potential Libraries:

- Consider using a library for sending emails (e.g., Nodemailer).
- A secure password hashing library like bcrypt or argon2 is recommended.

**6. How This Fits with Existing Work (Optional)**

- This implementation builds upon the previously developed work with the team..
- Ensure data models and API endpoints are compatible for a smooth integration.

### PR DESCRIPTION
**Pull request by maxCastro1**
###  Implement User Registration Feature

This pull request implements the user registration feature for the e-commerce platform. It allows users to sign up as either a vendor or a buyer and captures relevant information such as name, email, password, and any additional details required for each user type.

**What does this pull request do?**

- Implements a user registration API endpoint.
- Validates user input (e.g., email, password, user type).
- Checks for existing users (email uniqueness).
- Hashes and stores user passwords securely using bcrypt.
- Creates user records in the database with appropriate user types (vendor or buyer).
- Returns appropriate success or error responses.

### CODE REVIEW

File: src/routes/index.ts
Comment (faid-terence) [inline src/routes/index.ts:1]: the route should be `router.post('/register', registerUser)`
and set the auth routes in index.ts

**Review by faid-terence — CHANGES_REQUESTED** [review]
@maxCastro1 !
registration should be in `src/controllers/authController`

File: src/configs/db_config.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  host: process.env.DB_HOST || 'localhost',
+  port: parseInt(process.env.DB_PORT || '5432'),
+  username: process.env.DB_USERNAME,
+  password: process.env.DB_PASSWORD,
+  database: process.env.DB_DATABASE,
+
+  synchronize: process.env.NODE_ENV === 'dev' ? false : false,
+
+  logging: process.env.NODE_ENV === 'dev' ? false : false,
+  entities: [User],
```
Comment (elijahladdie) [inline src/configs/db_config.ts]: Why are you checking if node_env is dev and assign false to both states ?,
can you try to make that understandable or use different approach on this

File: src/controllers/index.ts [hunk 2]
```diff
@@ -1,5 +1,47 @@
+import { Request, Response } from 'express';
+import {OrmConfig} from '../configs/db_config';
+import { User } from '../entity/User';
+import bcrypt from 'bcrypt';
 // export all controllers
 function myFunction (): void {
   console.log('Hello');
 }
 myFunction();
```
Comment (elijahladdie) [inline src/controllers/index.ts]: If this function is doing nothing i thing it's better if you delete it

File: src/entity/User.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    @Column()
+    email!: string;
+  
+    @Column()
+    password!: string;
+  
+    @Column({ default: "User"} )
+    userType!: string;
+  }
+
```
Comment (aimedivin) [inline src/entity/User.ts]: @maxCastro1 Besides what we have, think about adding more info(fields) about users, such as accountStatus (to indicate whether it is active or inactive/blocked ), contact info (phone number, physical adrress),  etc...

**Review by aimedivin — CHANGES_REQUESTED** [review]

File: src/entity/User.ts (hunk 3 above)
Comment (faid-terence) [reply src/entity/User.ts]: @maxCastro1 
Please consider these feedbacks !

File: src/controllers/index.ts (hunk 2 above)
Comment (faid-terence) [reply src/controllers/index.ts]: @maxCastro1 !
Look into this as well !

File: src/configs/db_config.ts (hunk 1 above)
Comment (maxCastro1) [reply src/configs/db_config.ts]: this file was deleted it turns out it wasnt supposed to be here , we are using ormconfig.js now

File: src/controllers/index.ts (hunk 2 above)
Comment (maxCastro1) [reply src/controllers/index.ts]: noted

File: src/routes/index.ts
Comment (maxCastro1) [reply src/routes/index.ts:1]: corrected

File: src/entity/User.ts (hunk 3 above)
Comment (maxCastro1) [reply src/entity/User.ts]: yes i have added ,  the fields check and see if there any i missed

File: src/routes/index.ts [hunk 4]
```diff
@@ -1,14 +1,13 @@
...
-  res.send("Knights Ecommerce API");
-});
+
 router.get('/api/v1/status', (req: Request, res: Response) => {
   return responseSuccess(res, 202, 'This is a testing route that returns: 201');
 });
 
-// All routes should be imported here and get export after specifying first route
-// example router.use("/stock". stockRoutes) =>:: First import stockRoutes and use it here, This shows how the route export will be handled
+router.use('/user', userRoutes);
```
Comment (aimedivin) [inline src/routes/index.ts:12]: @maxCastro1  since you updated this API status route, don't forget to update its test to ensure our workflow job passes smoothly, also seems '/user' is not functional, it might be best to remove it.

**Review by aimedivin — CHANGES_REQUESTED** [review]

File: src/routes/index.ts (hunk 4 above)
Comment (maxCastro1) [reply src/routes/index.ts:12]: yes , i have updated the routes for tests

File: src/controllers/authController.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
+import { Request, Response } from 'express';
+import { User } from '../entities/User';;
+import bcrypt from 'bcrypt';
+import { getRepository } from 'typeorm';
+
+
+class UserController {
+    static registerUser = async (req: Request, res: Response) => {
+        const { firstName, lastName, email, password, gender, phoneNumber, userType, status, verified, photoUrl } = req.body;
```
Comment (Calebgisa72) [inline src/controllers/authController.ts:9]: Can you consider giving the verified an initial value of false in the user entity so that their cannot be a need of inputting it, you can also make it to status and give it an initial value pf active so that when creating a user, a user can be having verification false and status active without a need of putting that in the request.

**Review by Calebgisa72 — COMMENTED** [review]
Can you consider giving the verified an initial value of false in the user entity so that their cannot be a need of inputting it, you can also make it to status and give it an initial value pf active so that when creating a user, a user can be having verification false and status active without a need of putting that in the request.
